firmware/modes/programmer/at89/at89.c

```diff
@@ -0,0 +1,654 @@
+// pins need to be changed alongside the clock
+static inline void zif_clock_write(zif_bits_t op_template, zif_bits_t op_clk,
+                            unsigned int cycles
+                            )
+{
+    for(unsigned char i = 0; i <= cycles; i++) {
+        zif_write(op_template);
+        zif_write(op_clk);
+    }
+}
+
+static unsigned char read_byte(unsigned int addr)
+{
+    /* 
+     * AT89C51 Read Pinout:
+     * 
+     * Target   Dir     ZIF pin#    Programmer port
+     * ------------------------------------------------------------------------
+     * RST      <-      09          RJ4                     // (high)
+     * PSEN     <-      29          RD7                     // (low)
+     * PROG     <-      30          RG0                     // (high)
+     * VPP      <-      31          RJ0                     // (high)
+     * VCC      <-      40          Vdd_40
+     * P0.{0-7) ->      39-32       RB{6,5,4,3,2}, RJ{3,2,1}      // PGM Data
+     * P1.{0-7} <-      1-8         RC{5,4,3,2}, RJ{7,6}, RC{6,7} // Addr
+     * P2.{0-3} <-      21-24       RE{4,5,6,1}             // Addr (contd.)
+     * P2.6     <-      27          RD5                     // ctrl (low)
+     * P2.7     <-      28          RD6                     // ctrl (low)
+     * P3.4     ->      14          RD1                     // Busy
+     * P3.6     <-      16          RG1                     // ctrl (high)
+     * P3.7     <-      17          RE0                     // ctrl (high)
+     */
+    
+    // Set pin direction
+    zif_bits_t dir = {  0,
+                        0b00100000,   // Busy signal (14)
+                        0,
+                        0b10000000,   // p0.7 (32)
+                        0b01111111 }; // p0.{6-0} (33-39)  
+    dir_write(dir);
+    
+    // Set Vdd / GND pinout  
+    set_vdd(vdd);
+    set_gnd(gnd);
+    
+    // Set voltages
+    vdd_val(5); // 5.0 v - 5.2 v
+    
+    // Allocate an empty zifbits struct for reading pin state
+    zif_bits_t response    = { 0, 0, 0, 0, 0 };
+
+    // Base pin setting for reading
+    zif_bits_t read_base = { 0b00000000,
+                             0b10000001,   // 3.6 ctrl (16), RST (9)
+                             0b00000001,   // 3.7 ctrl (17) 
+                             0b01100000,   // VPP (31), PROG (30)
+                             0b00000000 };
+
+    zif_bits_t read_clk;
+
+    // Mask in the address bits to the appropriate pins
+    mask_addr(read_base, addr);
+
+    // Give the clock on/off states to zif_clock_write(..) and loop 48 cycles
+    clock_write(read_base, 48);
+
+    // Read the current pin state (to read in the requested byte)
+    zif_read(response);
+
+    // We're done with the byte. Turn off all outputs.
+    zif_write(zbits_null);
+    
+    return zif_to_data(response);
+}
+
+static void read(unsigned int addr, unsigned int range)
+{    
+    printf("%03X ", addr);
+
+    
+    if (!range) { range = 1; } else {com_println("");}

```

Comment:
1) Formatting in if and else block is inconsistent: spaces between {} and statement in if block but not in else block. That is "if (...) { a } else {b}" should be "if (...) { a } else { b }"

2) I'm not a huge fan of one line if statements, let alone more complex statements like this. if you don't have a strong preference I'd prefer them to be split over multiple lines

Reply:
I don't have a strong preference, I can fix that next time I hit the code.